File: dictionary/javascript/algorithms.js
/*
 * Classic algorithms: searching, sorting, graph traversal, and dynamic
 * programming solutions implemented cleanly and without dependencies.
 */

export function binarySearch(sorted, target) {
  let low = 0;
  let high = sorted.length - 1;
  while (low <= high) {
    const middle = Math.floor((low + high) / 2);
    if (sorted[middle] === target) return middle;
    if (sorted[middle] < target) low = middle + 1;
    else high = middle - 1;
  }
  return -1;
}

export function linearSearch(items, target) {
  for (let index = 0; index < items.length; index += 1) {
    if (items[index] === target) return index;
  }
  return -1;
}

export function mergeSorted(first, second) {
  const merged = [];
  let i = 0;
  let j = 0;
  while (i < first.length && j < second.length) {
    if (first[i] <= second[j]) merged.push(first[i++]);
    else merged.push(second[j++]);
  }
  return merged.concat(first.slice(i), second.slice(j));
}

export function quickSort(array) {
  if (array.length <= 1) return array;
  const pivot = array[Math.floor(array.length / 2)];
  const less = [];
  const equal = [];
  const greater = [];
  for (const value of array) {
    if (value < pivot) less.push(value);
    else if (value > pivot) greater.push(value);
    else equal.push(value);
  }
  return [...quickSort(less), ...equal, ...quickSort(greater)];
}

export function insertionSort(array) {
  const result = [...array];
  for (let index = 1; index < result.length; index += 1) {
    const value = result[index];
    let position = index - 1;
    while (position >= 0 && result[position] > value) {
      result[position + 1] = result[position];
      position -= 1;
    }
    result[position + 1] = value;
  }
  return result;
}

export function selectionSort(array) {
  const result = [...array];
  for (let index = 0; index < result.length - 1; index += 1) {
    let smallest = index;
    for (let cursor = index + 1; cursor < result.length; cursor += 1) {
      if (result[cursor] < result[smallest]) smallest = cursor;
    }
    if (smallest !== index) {
      [result[index], result[smallest]] = [result[smallest], result[index]];
    }
  }
  return result;
}

export function twoSum(nums, target) {
  const seen = new Map();
  for (let index = 0; index < nums.length; index += 1) {
    const complement = target - nums[index];
    if (seen.has(complement)) return [seen.get(complement), index];
    seen.set(nums[index], index);
  }
  return [];
}

export function longestCommonPrefix(strings) {
  if (strings.length === 0) return '';
  let prefix = strings[0];
  for (const value of strings.slice(1)) {
    while (!value.startsWith(prefix)) {
      prefix = prefix.slice(0, -1);
      if (prefix === '') return '';
    }
  }
  return prefix;
}

export function levenshteinDistance(first, second) {
  const rows = first.length + 1;
  const cols = second.length + 1;
  const table = Array.from({ length: rows }, () => new Array(cols).fill(0));
  for (let row = 0; row < rows; row += 1) table[row][0] = row;
  for (let col = 0; col < cols; col += 1) table[0][col] = col;
  for (let row = 1; row < rows; row += 1) {
    for (let col = 1; col < cols; col += 1) {
      const cost = first[row - 1] === second[col - 1] ? 0 : 1;
      table[row][col] = Math.min(
        table[row - 1][col] + 1,
        table[row][col - 1] + 1,
        table[row - 1][col - 1] + cost
      );
    }
  }
  return table[first.length][second.length];
}

export function isAnagram(first, second) {
  const normalize = (text) => text.toLowerCase().split('').sort().join('');
  return normalize(first) === normalize(second);
}

export function missingNumber(nums) {
  const length = nums.length;
  const expected = (length * (length + 1)) / 2;
  const actual = nums.reduce((total, value) => total + value, 0);
  return expected - actual;
}

export function majorityElement(nums) {
  let candidate = null;
  let count = 0;
  for (const value of nums) {
    if (count === 0) {
      candidate = value;
      count = 1;
    } else if (value === candidate) {
      count += 1;
    } else {
      count -= 1;
    }
  }
  return candidate;
}

export function maxSubarraySum(nums) {
  let best = nums[0];
  let current = nums[0];
  for (let index = 1; index < nums.length; index += 1) {
    current = Math.max(nums[index], current + nums[index]);
    best = Math.max(best, current);
  }
  return best;
}

export function mergeIntervals(intervals) {
  const ordered = [...intervals].sort((a, b) => a[0] - b[0]);
  const merged = [];
  for (const interval of ordered) {
    const last = merged[merged.length - 1];
    if (last && interval[0] <= last[1]) {
      last[1] = Math.max(last[1], interval[1]);
    } else {
      merged.push([interval[0], interval[1]]);
    }
  }
  return merged;
}

export function findDuplicates(nums) {
  const seen = new Set();
  const duplicates = new Set();
  for (const value of nums) {
    if (seen.has(value)) duplicates.add(value);
    seen.add(value);
  }
  return [...duplicates];
}

export function longestIncreasingSubsequence(nums) {
  const tails = [];
  for (const value of nums) {
    let low = 0;
    let high = tails.length;
    while (low < high) {
      const middle = Math.floor((low + high) / 2);
      if (tails[middle] < value) low = middle + 1;
      else high = middle;
    }
    tails[low] = value;
  }
  return tails.length;
}

export function coinChange(coins, amount) {
  const table = new Array(amount + 1).fill(Infinity);
  table[0] = 0;
  for (let total = 1; total <= amount; total += 1) {
    for (const coin of coins) {
      if (coin <= total) {
        table[total] = Math.min(table[total], table[total - coin] + 1);
      }
    }
  }
  return table[amount] === Infinity ? -1 : table[amount];
}

export function knapsack(weights, values, capacity) {
  const count = weights.length;
  const table = Array.from({ length: count + 1 }, () => new Array(capacity + 1).fill(0));
  for (let item = 1; item <= count; item += 1) {
    for (let weight = 1; weight <= capacity; weight += 1) {
      if (weights[item - 1] <= weight) {
        table[item][weight] = Math.max(
          table[item - 1][weight],
          table[item - 1][weight - weights[item - 1]] + values[item - 1]
        );
      } else {
        table[item][weight] = table[item - 1][weight];
      }
    }
  }
  return table[count][capacity];
}

export function breadthFirstSearch(graph, start) {
  const visited = new Set();
  const queue = [start];
  const order = [];
  while (queue.length > 0) {
    const node = queue.shift();
    if (visited.has(node)) continue;
    visited.add(node);
    order.push(node);
    for (const neighbor of graph[node] || []) {
      if (!visited.has(neighbor)) queue.push(neighbor);
    }
  }
  return order;
}

export function depthFirstSearch(graph, start) {
  const visited = new Set();
  const order = [];
  const visit = (node) => {
    if (visited.has(node)) return;
    visited.add(node);
    order.push(node);
    for (const neighbor of graph[node] || []) visit(neighbor);
  };
  visit(start);
  return order;
}

export function hasCycle(graph) {
  const visiting = new Set();
  const done = new Set();
  const visit = (node) => {
    if (done.has(node)) return false;
    if (visiting.has(node)) return true;
    visiting.add(node);
    for (const neighbor of graph[node] || []) {
      if (visit(neighbor)) return true;
    }
    visiting.delete(node);
    done.add(node);
    return false;
  };
  return Object.keys(graph).some((node) => visit(node));
}

export function topologicalSort(graph) {
  const visited = new Set();
  const order = [];
  const visit = (node) => {
    if (visited.has(node)) return;
    visited.add(node);
    for (const neighbor of graph[node] || []) visit(neighbor);
    order.push(node);
  };
  for (const node of Object.keys(graph)) visit(node);
  return order.reverse();
}

export function rotateMatrix(matrix) {
  return matrix[0].map((_, column) => matrix.map((row) => row[column]).reverse());
}

File: dictionary/lua/algorithms.lua
-- insertion_sort sorts a list in place by insertion.
local function insertion_sort(values)
	for i = 2, #values do
		local key = values[i]
		local j = i - 1
		while j >= 1 and values[j] > key do
			values[j + 1] = values[j]
			j = j - 1
		end
		values[j + 1] = key
	end
	return values
end

-- selection_sort repeatedly selects the smallest remaining element.
local function selection_sort(values)
	for i = 1, #values do
		local smallest = i
		for j = i + 1, #values do
			if values[j] < values[smallest] then
				smallest = j
			end
		end
		values[i], values[smallest] = values[smallest], values[i]
	end
	return values
end

-- bubble_sort bubbles the largest values to the end.
local function bubble_sort(values)
	local n = #values
	for i = 1, n do
		local swapped = false
		for j = 1, n - i do
			if values[j] > values[j + 1] then
				values[j], values[j + 1] = values[j + 1], values[j]
				swapped = true
			end
		end
		if not swapped then
			break
		end
	end
	return values
end

-- merge combines two sorted lists into one sorted list.
local function merge(left, right)
	local result = {}
	local i, j = 1, 1
	while i <= #left and j <= #right do
		if left[i] <= right[j] then
			result[#result + 1] = left[i]
			i = i + 1
		else
			result[#result + 1] = right[j]
			j = j + 1
		end
	end
	while i <= #left do
		result[#result + 1] = left[i]
		i = i + 1
	end
	while j <= #right do
		result[#result + 1] = right[j]
		j = j + 1
	end
	return result
end

-- merge_sort sorts a list recursively by divide and conquer.
local function merge_sort(values)
	if #values <= 1 then
		return values
	end
	local mid = math.floor(#values / 2)
	local left = {}
	local right = {}
	for i = 1, mid do
		left[i] = values[i]
	end
	for i = mid + 1, #values do
		right[i - mid] = values[i]
	end
	return merge(merge_sort(left), merge_sort(right))
end

-- binary_search finds a value in a sorted list in O(log n).
local function binary_search(values, target)
	local lo, hi = 1, #values
	while lo <= hi do
		local mid = math.floor((lo + hi) / 2)
		if values[mid] == target then
			return mid
		end
		if values[mid] < target then
			lo = mid + 1
		else
			hi = mid - 1
		end
	end
	return nil
end

-- linear_search scans an unsorted list for a value.
local function linear_search(values, target)
	for i, v in ipairs(values) do
		if v == target then
			return i
		end
	end
	return nil
end

-- two_sum finds indices of two values that add up to a target.
local function two_sum(values, target)
	local seen = {}
	for i, v in ipairs(values) do
		local needed = target - v
		if seen[needed] then
			return seen[needed], i
		end
		seen[v] = i
	end
	return nil
end

-- max_subarray finds the largest sum of a contiguous subarray.
local function max_subarray(values)
	if #values == 0 then
		return 0
	end
	local best = values[1]
	local current = values[1]
	for i = 2, #values do
		current = math.max(values[i], current + values[i])
		best = math.max(best, current)
	end
	return best
end

-- lis_length returns the length of the longest increasing subsequence.
local function lis_length(values)
	if #values == 0 then
		return 0
	end
	local lengths = {}
	local best = 1
	for i = 1, #values do
		lengths[i] = 1
		for j = 1, i - 1 do
			if values[j] < values[i] then
				lengths[i] = math.max(lengths[i], lengths[j] + 1)
			end
		end
		best = math.max(best, lengths[i])
	end
	return best
end

-- knapsack_01 computes the maximum value under a weight limit.
local function knapsack_01(weights, values, capacity)
	local dp = {}
	for w = 0, capacity do
		dp[w] = 0
	end
	for i = 1, #weights do
		for w = capacity, weights[i], -1 do
			dp[w] = math.max(dp[w], dp[w - weights[i]] + values[i])
		end
	end
	return dp[capacity]
end

-- edit_distance computes the Levenshtein distance of two strings.
local function edit_distance(a, b)
	local prev = {}
	for j = 0, #b do
		prev[j] = j
	end
	for i = 1, #a do
		local curr = { [0] = i }
		for j = 1, #b do
			local cost = a:sub(i, i) == b:sub(j, j) and 0 or 1
			curr[j] = math.min(prev[j] + 1, curr[j - 1] + 1, prev[j - 1] + cost)
		end
		prev = curr
	end
	return prev[#b]
end

-- lcs_length returns the length of the longest common subsequence.
local function lcs_length(a, b)
	local table = {}
	for i = 0, #a do
		table[i] = {}
		for j = 0, #b do
			table[i][j] = 0
		end
	end
	for i = 1, #a do
		for j = 1, #b do
			if a:sub(i, i) == b:sub(j, j) then
				table[i][j] = table[i - 1][j - 1] + 1
			else
				table[i][j] = math.max(table[i - 1][j], table[i][j - 1])
			end
		end
	end
	return table[#a][#b]
end

-- coin_change returns the fewest coins for a target amount.
local function coin_change(coins, amount)
	local dp = {}
	dp[0] = 0
	for i = 1, amount do
		dp[i] = math.huge
	end
	for i = 1, amount do
		for _, coin in ipairs(coins) do
			if coin <= i then
				dp[i] = math.min(dp[i], dp[i - coin] + 1)
			end
		end
	end
	if dp[amount] == math.huge then
		return nil
	end
	return dp[amount]
end

-- is_anagram checks whether two strings reuse the same letters.
local function is_anagram(a, b)
	if #a ~= #b then
		return false
	end
	local counts = {}
	for ch in a:gmatch(".") do
		counts[ch] = (counts[ch] or 0) + 1
	end
	for ch in b:gmatch(".") do
		counts[ch] = (counts[ch] or 0) - 1
		if counts[ch] < 0 then
			return false
		end
	end
	return true
end

-- majority_element finds the value appearing more than n/2 times.
local function majority_element(values)
	local candidate, balance = nil, 0
	for _, v in ipairs(values) do
		if balance == 0 then
			candidate = v
		end
		if v == candidate then
			balance = balance + 1
		else
			balance = balance - 1
		end
	end
	local count = 0
	for _, v in ipairs(values) do
		if v == candidate then
			count = count + 1
		end
	end
	if count > #values / 2 then
		return candidate
	end
	return nil
end

-- matrix_multiply multiplies two matrices with compatible dimensions.
local function matrix_multiply(a, b)
	local rows, inner, cols = #a, #a[1], #b[1]
	if inner ~= #b then
		return nil
	end
	local result = {}
	for i = 1, rows do
		result[i] = {}
		for j = 1, cols do
			local total = 0
			for k = 1, inner do
				total = total + a[i][k] * b[k][j]
			end
			result[i][j] = total
		end
	end
	return result
end

-- sieve returns the primes up to and including a limit.
local function sieve(limit)
	if limit < 2 then
		return {}
	end
	local composite = {}
	for i = 2, math.floor(math.sqrt(limit)) do
		if not composite[i] then
			for j = i * i, limit, i do
				composite[j] = true
			end
		end
	end
	local primes = {}
	for i = 2, limit do
		if not composite[i] then
			primes[#primes + 1] = i
		end
	end
	return primes
end

-- floyd_cycle detects a cycle in a functional graph of indices.
local function floyd_cycle(next)
	if #next < 2 then
		return false
	end
	local slow, fast = 1, 1
	for _ = 1, #next + 1 do
		slow = next[slow]
		fast = next[fast]
		if not fast then
			return false
		end
		fast = next[fast]
		if slow == fast then
			return true
		end
	end
	return false
end

-- count_inversions counts pairs out of order (naive O(n^2)).
local function count_inversions(values)
	local count = 0
	for i = 1, #values do
		for j = i + 1, #values do
			if values[i] > values[j] then
				count = count + 1
			end
		end
	end
	return count
end

-- partition_qs reorders a list around a pivot and returns its index.
local function partition_qs(values, lo, hi)
	local pivot = values[hi]
	local i = lo - 1
	for j = lo, hi - 1 do
		if values[j] <= pivot then
			i = i + 1
			values[i], values[j] = values[j], values[i]
		end
	end
	values[i + 1], values[hi] = values[hi], values[i + 1]
	return i + 1
end
